xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -504,9 +533,69 @@ private <T extends ResourceUpdate> void handleResourceUpdate(
     }
   }
 
-  /**
-   * Tracks a single subscribed resource.
-   */
+  @Override
+  public SettableFuture<Void> reportServerConnections(ServerConnectionCallback callback) {
+    SettableFuture<Void> future = SettableFuture.create();
+    syncContext.execute(() -> {
+      serverCpClientMap.forEach((serverInfo, controlPlaneClient) ->
+          callback.reportServerConnectionGauge(
+              controlPlaneClient.hasWorkingAdsStream() ? 1 : 0,
+              target,
+              serverInfo.target()));
+    });
+    future.set(null);
+    return future;
+  }
+
+  @Override
+  public SettableFuture<Void> reportResourceCounts(ResourceCallback callback) {
+    SettableFuture<Void> future = SettableFuture.create();
+    syncContext.execute(() -> {
+      Map<XdsResourceType<?>, Map<String, Long>> resourceCountsByType =
+          getResourceCountsByType();
+      reportResourceCountsToCallback(callback, resourceCountsByType);
+    });
+    future.set(null);
+    return future;
+  }
+
+  private String cacheStateFromResourceStatus(ResourceMetadata metadata, boolean isResourceCached) {
+    String status = metadata.getStatus().toString().toLowerCase(Locale.ROOT);
+    return metadata.getStatus() == ResourceMetadataStatus.NACKED && isResourceCached
+        ? status + "_but_cached" : status;
+  }
+
+  /** Calculates number of resources by ResourceType and ResourceSubscriber.metadata.status. */
+  Map<XdsResourceType<?>, Map<String, Long>> getResourceCountsByType() {
+    Map<XdsResourceType<?>, Map<String, Long>> resourceCountsByType = new HashMap<>();
```

Comment:
This map is only written to in this method, so it isn't actually necessary. If `ResourceCallback` was an argument to this method:
```java
// instead of:
      resourceCountsByType.put(resourceType, resourceCountsByState);
// you could do:
      resourceCountsByState.forEach((cacheState, count) ->
          callback.reportResourceCountGauge(count, cacheState, resourceType.typeUrl(), target));
```

There's not many XdsResourceType nor cache_state, so creating the full map isn't that much memory. Most of the cost right now is `toLowerCase()` and `Long` object creation for each entry. And optimizing those has no real impact to the memory cost of the data structure you return.

Although it looks like this is going to have to be reorganized when `authority` is added, that still won't add much additional memory.

But to really benefit from the callback, we probably shouldn't do any aggregation within XdsClientImpl. Instead, call the callback with the raw data and have it aggregate as it needs. That means passing `ResourceMetadataStatus` and `isResourceCached` separately and letting it determine the string to use. That makes the implementation here less sensitive to the whims of the metric reporter.

Reply:
I initially separated to into multiple methods for testing the logic of aggregating resource counts separately, but instead tested using callback itself. Removed `getResourceCountsByType()` and `reportResourceCountsToCallback()`. 

> Instead, call the callback with the raw data and have it aggregate as it needs.
Will make this change after populating `subscriber.data` value in `ResourceMetadata`.